type-inference-in-readr-and-arrow.rmd

```diff
@@ -0,0 +1,271 @@
+# Type inference in readr and arrow
+
+The CSV format is widely used in data science, and at its best works
+well as a simple human-readable format that is widely known and
+understood. The simplicity of CSVs though, as a basic text format also
+has its drawbacks.  One is that it contains no information about data
+types of its columns, and if you're working with CSVs in an application
+more complex than a text editor, those data types must be inferred by
+whatever is reading the data.
+
+In this blog post, I'm going to discuss how CSV type inference works in
+the R packages [readr](https://readr.tidyverse.org/) and [arrow](https://arrow.apache.org/docs/r/), and highlight the differences between
+the two.
+
+<p align="center">
+![](images/readr_hex.png){width=200px}
+![](images/arrow-logo_hex_black-txt_white-bg.png){width=200px}
+</p>
+
+```{r, include=FALSE}
+library(magrittr)
+library(arrow)
+library(readr)
+```
+
+## How does type inference work in readr?
+
+In readr, the first step is reading in a subset of the data (by default,
+1000 rows, though this can be customised). The type inference is done by
+a C++ function in readr called `collectorGuess()` which guesses types in
+the following order:
+
+-   Does the column contain 0 rows? If yes, return "character"
+
+-   Are all values missing? If yes, return "logical"
+
+-   Tries to parse column to each of these formats and returns the first
+    one it successfully parses:
+
+    -   Logical
+
+    -   Integer (though the default is to not look for these)
+
+    -   Double
+
+    -   Number (a special type which can remove characters from strings
```

Comment:
Nitpicky: I try to avoid explicitly talking about "Number" as a *type* that's distinct from integer or double. `col_number()` will return double. The special-ness of `col_number()` is in the parsing/guessing strategy, not the return value.

Reply:
Not nitpicky at all, it's a fair comment! Updated now.